### ISSUE TITLE
Deterministic output order

### ISSUE THREAD

**Issue by andys8** [rtfeldman/elm-css#488]
### Issue

Having conflicting properties can lead to non-deterministic results. `min-width` is set twice. This will have the same css specificity. This is most likely a mistake of the user. The issue is: The result looked fine in dev mode (with `create-elm-app`), but was broken in production. The precedence changed.

#### Dev
![image](https://user-images.githubusercontent.com/13085980/49081010-b3eb2f00-f245-11e8-98f6-9a4f9f66caa1.png)

#### Prod
![image](https://user-images.githubusercontent.com/13085980/49081188-44c20a80-f246-11e8-993c-d63c53740e41.png)

https://css-tricks.com/precedence-css-order-css-matters/

### Solution

Since the names seem to be deterministic, I think it is the order the classes are printed into the resulting styles. If this is the case, sorting them by name would work.

**Comment by rtfeldman** [comment]
Interesting!

Sorting the classes before emitting would presumably address this, but at a performance cost. We could also de-dupe the styles by storing them in a dictionary rather than a list, also at a performance cost.

What I'm wondering is: why does the output order differ between development and production? I can't think of a reason why that should be true. The only thing I can think of is https://github.com/elm/compiler/issues/1836 maybe?

**Comment by andys8** [comment]
I was using `create-elm-app`. So there are differences between `start` and `build`: Different Webpack configuration, hot module reloading (?), `--debug`, `--optimize`.

The issue you mentioned is highly interesting. If `--optimize` would be the cause, sorting lists wouldn't make a difference. For Evan on the other side it's probably surprising someone would depend on the order of style definitions.

**Comment by rtfeldman** [comment]
Hm, what happens if you keep everything else about the configuration the same, but remove `--optimize`? Is it still reproducible then?

**Comment by andys8** [comment]
I'm trying to limit the cause, but I'm currently not able to reproduce it. As of now I can't be sure, that the identical code was used for the example above.

As you can see the class names were the same. Can you give me insights what code changes could change class name order?

**Comment by rtfeldman** [comment]
> Can you give me insights what code changes could change class name order?

That's the thing - I literally can't think of any. 😅 

The generated class name is a pure function of the `Style` values involved, and the output is determined by iterating through the virtual dom nodes, concatenating all their styles, and emitting that to a `<style>` element.

The `elm-css` code involved should be exactly the same in development versus production, and I'm not aware of any tools that can affect that.

In general, it should be true that any Elm code built in development and in production should do exactly the same thing in both environments—the only different should be that code built with `--optimize` should be smaller and run faster. That's why a compiler bug with `--optimize` is the only potential culprit I can think of.

**Comment by andys8** [comment]
Yeah, I'm sorry, as long as I can't reproduce it, in a useful way, it doesn't make a lot of sense to investigate.

I noticed, the styles list depends on what is actually shown. The diffs between the styles of multiple pages seem to just add and remove classes. Is there a merge strategy when the multiple elements use the same "class" (inline styles)?

So let's have this imaginary example: Can you think of the scenario, that there are already two conflicting styles resulting on two classes. And there is a code change somewhere, which triggers a different order or output? Can you think of situations?

**Comment by rtfeldman** [comment]
> Is there a merge strategy when the multiple elements use the same "class" (inline styles)?

Sorry, I don't quite follow - why would a merge be necessary?

`elm-css` should generate exactly one `class` per element, no matter what. If two elements have the same `class`, that means their styles are identical.

**Comment by andys8** [comment]
What about the issue https://github.com/rtfeldman/elm-css/issues/415? It has a 0.18 example, but sounds pretty much like the "multi class" issue, we have here. SSCCE of that issue: https://ellie-app.com/9P7wNngFQa1/1

**Comment by andys8** [comment]
Minimal example for two classes on one element:

```elm
module Main exposing (main)

import Browser
import Css exposing (..)
import Html.Styled as Html exposing (Html, button, text, toUnstyled)
import Html.Styled.Attributes exposing (css)


button1 =
    Html.button
        [ css [ cursor pointer ]
        , css [ border zero ]
        ]
        [ text "button" ]


main =
    button1 |> toUnstyled
```
https://ellie-app.com/429JgfpfjZsa1

![image](https://user-images.githubusercontent.com/13085980/49146768-279f4180-f303-11e8-94f6-0892201ae7ac.png)

**Comment by pdamoc** [comment]
> Minimal example for two classes on one element:

 wrapping the  code in a div solves this problem 

`div [] [ button1 ] |> toUnstyled`
